<launch>
  <!--
    ###################
    ## Configuration ##
    ###################
  -->
  <arg name='host' default='localhost'/>
  <arg name='port' default='2000'/>
  <arg name='town' default='Town02'/>

  <arg name='agent' default='agent'/>

  <arg name='role_name' default='ego_vehicle'/>
  <!-- use comma separated format "x,y,z,roll,pÃ¬tch,yaw" -->
  <arg name='spawn_point' default='193.78,218.79,0.5,0.0,0.0,90.0'/>
  <arg name='synchronous_mode' default='True'/>

  <arg name='sensor_definition_file' default='$(find carla_autoware_agent)/$(arg agent)/config/sensors.json'/>

  <!-- vehicle info -->
  <arg name='vehicle_model' default='vehicle.toyota.prius'/>
  <arg name='vehicle_length' default='4.54'/>
  <arg name='vehicle_width' default='2.00'/>
  <arg name='vehicle_wheelbase' default='2.7'/>

  <arg name='use_ground_truth_localization' default='false'/>
  <arg name='use_ground_truth_detections' default='false'/>
  <arg name='use_manual_control' default='false'/>

  <!--
    ###########################
    ## CARLA Autoware bridge ##
    ###########################
  -->
  <include file='$(find carla_autoware_bridge)/launch/carla_autoware_bridge.launch'>
    <arg name='host' value='$(arg host)'/>
    <arg name='port' value='$(arg port)'/>
    <arg name='town' value='$(arg town)'/>
    <arg name='vehicle_filter' value='$(arg vehicle_model)'/>
    <arg name='role_name' value='$(arg role_name)'/>
    <arg name='spawn_point' value='$(arg spawn_point)'/>
    <arg name='sensor_definition_file' value='$(arg sensor_definition_file)'/>
    <arg name='use_manual_control' value='$(arg use_manual_control)'/>
  </include>

  <!--
    ##################
    ## Agent bridge ##
    ##################
  -->
  <include file='$(find carla_autoware_agent)/$(arg agent)/bridge.launch'>
    <arg name='role_name' value='$(arg role_name)'/>
    <arg name='wheelbase' value='$(arg vehicle_wheelbase)'/>
  </include>

  <!--
    ####################
    ## Autoware agent ## 
    ####################
  -->
  <include file='$(find carla_autoware_agent)/$(arg agent)/my_agent.launch'>
    <arg name='town' value='$(arg town)'/>
    <arg name='vehicle_length' value='$(arg vehicle_length)'/>
    <arg name='vehicle_width' value='$(arg vehicle_width)'/>
    <arg name='use_ground_truth_localization' value='$(arg use_ground_truth_localization)'/>
    <arg name='use_ground_truth_detections' value='$(arg use_ground_truth_detections)'/>
    <arg name='use_manual_control' value='$(arg use_manual_control)'/>
  </include>

  
  <!--
    ####################
    ## Image error    ## 
    ####################
  -->  
  <include file='$(find image_error_injection)/launch/image_error_injection.launch'>
  </include>
  
  <include file='$(find image_error_injection)/launch/spawn_objects.launch'>
  </include>
        
        
  <!--
    ###################
    ## Visualization ## 
    ###################
  -->
  <arg name='rvizconfig' default='$(find carla_autoware_agent)/$(arg agent)/rviz/config.rviz'/>
  <node name='rviz' pkg='rviz' type='rviz' args='-d $(arg rvizconfig)'/>

  <!--
    ####################
    ## set goal       ##
    ####################
  -->
  <arg name="my_goal" default="'{header: {seq: 0, stamp: {secs: 0, nsecs: 0}, frame_id: ''}, pose: {position: {x: 193.78, y: -118.79, z: 0.5}, orientation:{x: 0.0, y: 0.0, z: 0.7071, w: 0.7071}}}'"/>
  <node pkg='rostopic' type='rostopic' name='set_goal'
        args="pub /move_base_simple/goal geometry_msgs/PoseStamped $(arg my_goal)"/>  
        
  <!--
    ####################
    ## rosbag       ##
    ####################
  -->
  <node pkg="rosbag" type="record" name="rosbag_noisefree"
       args="record -O bag_noise_11182021 -e /ground_truth_bbox /detection/(.*)/objects /image_(.*) /carla/ego_vehicle/(.*)"/>


</launch>

